File: backend/app/risk_rules.py
def aip_risk(aip):
    if aip is None:
        return "More Data Needed", "grey"
    if aip <= 0.10:
        return "Low", "green"
    if aip <= 0.24:
        return "Moderate", "yellow"
    return "High", "red"


def apri_risk(apri):
    if apri is None:
        return "More Data Needed", "grey"
    if apri <= 0.5:
        return "Low", "green"
    if apri <= 1.5:
        return "Moderate", "yellow"
    return "High", "red"


def fib4_risk(fib4):
    if fib4 is None:
        return "More Data Needed", "grey"
    if fib4 < 1.3:
        return "Low", "green"
    if fib4 <= 2.67:
        return "Moderate", "yellow"
    return "High", "red"


def tyg_risk(tyg):
    if tyg is None:
        return "More Data Needed", "grey"
    if tyg <= 4.5:
        return "Low", "green"
    return "High", "red"


def nlr_risk(nlr):
    if nlr is None:
        return "More Data Needed", "grey"
    if nlr <= 3:
        return "Low", "green"
    if nlr <= 9:
        return "Moderate", "yellow"
    return "High", "red"


def spO2_risk(spo2):
    if spo2 is None:
        return "More Data Needed", "grey"
    if spo2 >= 95:
        return "Low concern", "green"
    if 90 <= spo2 < 95:
        return "Moderate monitoring suggested", "yellow"
    return "High attention needed", "red"


def lar_rule(lar):
    if lar is None:
        return "More Data Needed", "grey"
    if lar < 3:
        return "Low", "green"
    return "Moderate", "yellow"


def egfr_rule(egfr):
    if egfr is None:
        return "More Data Needed", "grey"
    if egfr >= 90:
        return "Low concern", "green"
    if egfr >= 60:
        return "Moderate monitoring suggested", "yellow"
    return "High attention needed / clinical review suggested", "red"


def fli_risk(fli):
    if fli is None:
        return "More Data Needed", "grey"
    if fli < 30:
        return "Low", "green"
    if fli <= 60:
        return "Moderate", "yellow"
    return "High", "red"


def nafld_risk(score):
    if score is None:
        return "More Data Needed", "grey"
    if score < -1.455:
        return "Low", "green"
    if score <= 0.676:
        return "Moderate", "yellow"
    return "High", "red"


def metabolic_risk(fasting=None, hba1c=None, ppbs=None, random_blood_sugar=None):
    values = [v for v in [fasting, hba1c, ppbs, random_blood_sugar] if v is not None]
    if not values:
        return "More Data Needed", "grey"
    high = (
        (fasting is not None and fasting >= 126)
        or (hba1c is not None and hba1c >= 6.5)
        or (ppbs is not None and ppbs >= 200)
        or (random_blood_sugar is not None and random_blood_sugar >= 200)
    )
    moderate = (
        (fasting is not None and fasting >= 100)
        or (hba1c is not None and hba1c >= 5.7)
        or (ppbs is not None and ppbs >= 140)
        or (random_blood_sugar is not None and random_blood_sugar >= 140)
    )
    if high:
        return "High", "red"
    if moderate:
        return "Moderate", "yellow"
    return "Low", "green"


def tumor_marker_risk(value, low_cutoff):
    if value is None:
        return "More Data Needed", "grey"
    if value <= low_cutoff:
        return "Low / within awareness threshold", "green"
    return "High awareness indicator; clinical review suggested", "red"

File: frontend/lib/core/risk_rules.dart
class RiskRuleResult {
  const RiskRuleResult(this.level, this.color);

  final String level;
  final String color;
}

const lowConcern = 'Low Concern';
const monitor = 'Monitor';
const attentionNeeded = 'Attention Needed';
const moreDataNeeded = 'More Data Needed';

RiskRuleResult _low() => const RiskRuleResult(lowConcern, 'soft_green');
RiskRuleResult _monitor() => const RiskRuleResult(monitor, 'soft_amber');
RiskRuleResult _attention() =>
    const RiskRuleResult(attentionNeeded, 'soft_coral');
RiskRuleResult _more() => const RiskRuleResult(moreDataNeeded, 'soft_lavender');

RiskRuleResult aipRisk(double? value) {
  if (value == null) return _more();
  if (value <= 0.10) return _low();
  if (value <= 0.24) return _monitor();
  return _attention();
}

RiskRuleResult apriRisk(double? value) {
  if (value == null) return _more();
  if (value <= 0.5) return _low();
  if (value <= 1.5) return _monitor();
  return _attention();
}

RiskRuleResult fib4Risk(double? value) {
  if (value == null) return _more();
  if (value < 1.3) return _low();
  if (value <= 2.67) return _monitor();
  return _attention();
}

RiskRuleResult tygRisk(double? value) {
  if (value == null) return _more();
  if (value <= 4.5) return _low();
  return _attention();
}

RiskRuleResult nlrRisk(double? value) {
  if (value == null) return _more();
  if (value <= 3) return _low();
  if (value <= 9) return _monitor();
  return _attention();
}

RiskRuleResult spo2Risk(double? value) {
  if (value == null) return _more();
  if (value >= 95) return _low();
  if (value >= 90) return _monitor();
  return _attention();
}

RiskRuleResult egfrRisk(double? value) {
  if (value == null) return _more();
  if (value >= 90) return _low();
  if (value >= 60) return _monitor();
  return _attention();
}

RiskRuleResult tumorMarkerRisk(double? value, double cutoff) {
  if (value == null) return _more();
  if (value <= cutoff) return _low();
  return _attention();
}

RiskRuleResult larRisk(double? value) {
  if (value == null) return _more();
  if (value < 3) return _low();
  return _monitor();
}

RiskRuleResult fliRisk(double? value) {
  if (value == null) return _more();
  if (value < 30) return _low();
  if (value <= 60) return _monitor();
  return _attention();
}

RiskRuleResult nafldRisk(double? value) {
  if (value == null) return _more();
  if (value < -1.455) return _low();
  if (value <= 0.676) return _monitor();
  return _attention();
}

RiskRuleResult metabolicRisk({
  double? fasting,
  double? hba1c,
  double? ppbs,
  double? randomBloodSugar,
}) {
  final hasValue =
      [fasting, hba1c, ppbs, randomBloodSugar].any((value) => value != null);
  if (!hasValue) return _more();
  final high = (fasting != null && fasting >= 126) ||
      (hba1c != null && hba1c >= 6.5) ||
      (ppbs != null && ppbs >= 200) ||
      (randomBloodSugar != null && randomBloodSugar >= 200);
  if (high) return _attention();
  final moderate = (fasting != null && fasting >= 100) ||
      (hba1c != null && hba1c >= 5.7) ||
      (ppbs != null && ppbs >= 140) ||
      (randomBloodSugar != null && randomBloodSugar >= 140);
  if (moderate) return _monitor();
  return _low();
}

int severityRank(String? level) {
  final value = (level ?? '').toLowerCase();
  if (value.contains('attention') || value.startsWith('high')) return 3;
  if (value.contains('monitor') || value.startsWith('moderate')) return 2;
  if (value.contains('low') ||
      value.contains('optimal') ||
      value.contains('within awareness threshold')) {
    return 1;
  }
  return 0;
}

String overallRisk(Iterable<String?> levels) {
  final topRank = levels.fold<int>(
    0,
    (top, level) {
      final rank = severityRank(level);
      return rank > top ? rank : top;
    },
  );
  if (topRank >= 3) return attentionNeeded;
  if (topRank == 2) return monitor;
  if (topRank == 1) return lowConcern;
  return moreDataNeeded;
}
